scripts/new_upstream_snapshot.py

```diff
@@ -0,0 +1,627 @@
+        return cls(
+            changelog_details,
+            devel_distro,
+            changelog_distro,
+            head,
+        )
+
+
+def remove_line_from_file(filename, text):
+    """Utility function to remove a single line from a file."""
+    with open(filename, "r") as f:
+        lines = f.readlines()
+    with open(filename, "w") as f:
+        for line in lines:
+            if line.strip() != text.strip():
+                f.write(line)
+
+
+def merge_commitish(to: str) -> None:
+    """Perform 'git merge <commitish>' along with some error checking."""
+    try:
+        ref_name = capture(f"git describe --abbrev=8 {to}").stdout.strip()
+    except CalledProcessError as e:
+        raise RuntimeError(
+            f"'git describe' failed for {to}. "
+            "Is it a valid commtish or annotated tag?"
+        ) from e
+
+    command = f'git merge {to} -m "merge from {to} at {ref_name}"'
+    print(f"Running: {command}")
+    sh(command)
+
+
+def drop_cpicks(commitish):
+    """Drop any cpick files in d/p that we've pulled in from main.
+
+    Cpick files have their commit specified in the filename.
+    """
+    print("Dropping any cpicks that we've pulled in from main")
+    dropped_cpicks = []
+    for cpick in Path("debian/patches").glob("cpick*"):
+        hash = cpick.name.split("-")[1]
+        is_ancestor = (
+            sh(
+                f"git merge-base --is-ancestor {hash} {commitish}", check=False
+            ).returncode
+            == 0
+        )
+        if is_ancestor:
+            print(
+                f"Dropping file {cpick.name} as it is contained in the "
+                "upstream snapshot"
+            )
+            dropped_cpicks.append(cpick.name)
+            remove_line_from_file("debian/patches/series", cpick.name)
+            cpick.unlink()
+    if dropped_cpicks:
+        commit_msg = (
+            f"drop cherry picks included in {commitish}.\n\n"
+            "drop the following cherry picks:\n" + "\n".join(dropped_cpicks)
+        )
+        sh(f"git add debian/patches && git commit -m '{commit_msg}'")
+
+
+def refresh_patches(commitish):
+    """Refresh any non-cpick quilt patches.
+
+    For every quilt patch run:
+      quilt next
+      quilt push
+      quilt refresh
+
+    Changes from refresh will automatically be committed without prompting.
+
+    If automatic refresh fails, the script will exit with failure. It
+    is up to the user to manually fix the quilt patches and then rerun
+    the script with the '--post-patch' argument.
+    """
+    print("Attempting to automatically refresh quilt patches")
+    try:
+        while sh(f"{QUILT_COMMAND} next", check=False).returncode == 0:
+            sh(f"{QUILT_COMMAND} push")
+            sh(f"{QUILT_COMMAND} refresh", env=QUILT_ENV)
+    except CalledProcessError as e:
+        failed_patch = capture(f"{QUILT_COMMAND} next").stdout.strip()
+        raise RuntimeError(
```

Comment:
These tracebacks are busy when we barf the the cmdline. 
Objections to just capturing and printing the string of the error?
```diff
@@ -617,11 +641,15 @@ def parse_args() -> argparse.Namespace:
 
 if __name__ == "__main__":
     args = parse_args()
-    new_upstream_snapshot(
-        args.commitish,
-        args.bug,
-        args.first_devel_upload,
-        args.no_sru_bug,
-        args.first_sru,
-        args.post_patch,
-    )
+    try:
+        new_upstream_snapshot(
+            args.commitish,
+            args.bug,
+            args.first_devel_upload,
+            args.no_sru_bug,
+            args.first_sru,
+            args.post_patch,
+        )
+    except Exception as e:
+         print("%s" % e)
+         sys.exit(1)
```

Reply:
This isn't really an "end user" tool, so I thought it'd be better to know where exactly something failed because so often failures in scripts like this are the fault of the script, not the user. But I see your point. I'll add a custom exception for "expected" exceptions that will just print the text. Unexpected failures I want the traceback though.